src/poiFilter.ts

```diff
@@ -0,0 +1,120 @@
+import mapboxgl from 'mapbox-gl';
+
+interface Options {
+  filter?: Array<any>;
+  include?: boolean;
+}
+
+/**
+ * Select and switch 3D building.
+ */
+export function poiFilter(map: mapboxgl.Map, options?: Options) {
+  if (options?.filter) {
+    applyFilter(map, options?.filter, options?.include);
+  }
+
+  return {
+    setExcludeFilter(filter: Array<any>) {
+      applyFilter(map, filter, false);
+    },
+    setIncludeFilter(filter: Array<any>) {
+      applyFilter(map, filter, true);
+    },
+    unsetFilter() {
+      unsetFilter(map);
+    },
+  };
+}
+
+const poiLayers = [
+  'poi-level-1',
+  'poi-level-1-icon',
+  'poi-level-2',
+  'poi-level-2-icon',
+  'poi-level-3',
+  'poi-level-3-icon',
+];
+
+const filterId = ['==', ['literal', '__poiFilter'], '__poiFilter'];
+
+function pruneFilter(filter: any): any {
+  if (filter[0] !== 'all') {
+    return undefined;
+  }
+
+  return filter.filter(
+    (close: any) =>
+      !(
+        close instanceof Array &&
+        close.length >= 2 &&
+        close[0] === 'all' &&
+        JSON.stringify(close[1]) === JSON.stringify(filterId)
+      ),
+  );
+}
+
+function createFilterExpression(filter: Array<any>): Array<any> {
```

Comment:
I don't understand what kind of data `filter` could contain. Can you give me some examples please? I would like to avoid `any` as much as possible.

Reply:
It is a Mapbox gl expression https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/
I did not found the definition again, but was Array<any>.
It look like this https://github.com/teritorio/teritorio-tourism-gl-style/blob/master/style.json#L3232-L3238

Reply:
I'm talking about the `any` type and not the string `"any"` from Mapbox GL.

As I can see in the example code you gave me, we could use: `string | string[]`.

Do you mind if I add a commit in your PR?

Reply:
I'm also talking about `any` of TS.

But we need a recursive `string | string[]` and also integer, boolean, null, and array these.

https://github.com/teritorio/teritorio-tourism-gl-style/blob/master/style.json#L3244-L3250
Here, the filter is simple, but can be more complex, with deeper structures.

https://github.com/teritorio/teritorio-tourism-gl-style/blob/master/style.json#L3252-L3280
This one is also an expression.

Reply:
Ok, I just pushed a commit with a new custom type:

```ts
type MapboxValues = undefined | string | number | boolean | null | MapboxValues[];
```

If it's ok for you we can merge 👌

Reply:
Not sure.

Is type checked on runtime on browser? I think it is lock type in a way that its not required and add potential bug.

Reply:
No runtime check no, and even potentially incomplete, the new `MapboxValues` type helps to see some bugs.

For instance, you used `f.length`, thinking about `arrays`, but it also works with strings. The type showed that the following `map` and `join` calls were invalid on strings.

Reply:
ok